File: scripts/cluster_by_distances.js
var d3 = require('d3');
var topojson = require("topojson");

function proximityCluster(points, distance,dist) {
    //cluster points such that whenever two points are
    //within distance of each other, they are placed in
    //the same cluster

    //take m n-d points as input and
    //return an array of length m containing integers
    //where each integers encodes the cluster that the ith
    //point is in
    //
    // dist is a function that computes the distance between
    // any two points

    var clusters = [], links = [];
    var linksDict = {};
    
    // all points are in their own cluster
    for (var i = 0; i < points.length; i++) {
        clusters.push(i);
        linksDict[i] = [];
    }

    // sort the points by x values
    points = points.sort(function(a,b) { return a[0] - b[0]; });

    for (i = 0; i < points.length; i++) {
        //check for points to the right
        for (j = i+1; j < points.length; j++) {
            if (Math.abs(points[j][0] - points[i][0]) > distance)
                break;  // gone too far

            if (dist(points[i], points[j]) < distance) {
                // the two points are close enough to be linked
                links.push([i,j]);
                linksDict[i].push(j);
                linksDict[j].push(i);
            }
        }
    }

    var visited = d3.set();

    // traverse the list of links and assign clusters
    function dfs(currNode, prevNode) {
        visited.add(currNode);
        clusters[currNode] = Math.min(clusters[currNode], clusters[prevNode]);

        for (var i = 0; i < linksDict[currNode].length; i++) {
            if (visited.has(linksDict[currNode][i]))
                continue;

            dfs(linksDict[currNode][i], currNode);
        }
    }

    // go through every point and traverse its list of links
    // assigning clusters
    for (i = 0; i < points.length; i++) {
        if (visited.has(i))
            continue;

        dfs(i, i);
    }

    return {"clusters": clusters, "links": links};
}

function haversine(lat1, lon1, lat2, lon2) {
    /*
    Calculate the haversine distance between two points on the
    globe. Code taken from:

    http://stackoverflow.com/a/14561433
   */
    Number.prototype.toRad = function() {
        return this * Math.PI / 180;
    };

    var R = 6371; // km 
    //has a problem with the .toRad() method below.
    var x1 = lat2-lat1;
    var dLat = x1.toRad();  
    var x2 = lon2-lon1;
    var dLon = x2.toRad();  

    var a = Math.sin(dLat/2) * Math.sin(dLat/2) + 
                    Math.cos(lat1.toRad()) * Math.cos(lat2.toRad()) * 
                    Math.sin(dLon/2) * Math.sin(dLon/2);  
    var c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a)); 
    var d = R * c;

    return d;
}

function distanceBetweenNodes(node1, node2) { 
    return haversine(+node1['$'].lat, +node1['$'].lon, +node2['$'].lat, +node2['$'].lon);
}

function clusterOsmPoints(osmData, distance) {
    /* Take the input XML data structure and cluster
     * the points according to how far from each other they
     * are.
     *
     * Return a dictionary where the cluster names are the keys
     * and the values are lists of (lat, lon) pairs
     */
    var clusters = proximityCluster(osmData.osm.node, distance, distanceBetweenNodes);
    var clusterPoints = {};

    for (var i = 0; i < clusters.clusters.length; i++) {
        // iterate over every cluster assignment for each point
        var currCluster;

        // check if we've already seen this cluster of points before
        if (clusters.clusters[i] in clusterPoints) 
            currCluster = clusterPoints[clusters.clusters[i]]; 
        else
            currCluster = [];  // we haven't create a new one

        currCluster.push([+osmData.osm.node[i].$.lat, +osmData.osm.node[i].$.lon]);
        clusterPoints[clusters.clusters[i]] = currCluster;
    }

    return clusterPoints;
}

function clustersToTriangles(clusterPoints, distance) {
    /* Create alpha shapes for all of the clusters in clusterPoints
     *
     * Return a dictionary of the cells.
     */
    var newCellsDict = {};
    for (var clusterId in clusterPoints) {
        if (clusterPoints[clusterId].length < 2)
            continue;

        if (clusterPoints[clusterId].length == 2) {
            newCellsDict[clusterId] = [clusterPoints[clusterId]];
            continue;
        }

        var mesh = d3.geom.delaunay(clusterPoints[clusterId]).filter(function(t) {
            return  (haversine(t[0][0], t[0][1], t[1][0], t[1][1]) < distance &&
                     haversine(t[0][0], t[0][1], t[2][0], t[2][1]) < distance &&
                     haversine(t[2][0], t[2][1], t[1][0], t[1][1]) < distance);
        });
        
        newCellsDict[clusterId] = mesh;
    }

    return newCellsDict;
}

function clustersToAlphaShapes1(clusterPoints, distance) {
    /* Create alpha shapes for all of the clusters in clusterPoints
     *
     * Return a dictionary of the cells.
     */
    if (arguments.length == 1)
        distance = 1;

    var cellsDict = {};
    var newCellsDict = {};
    for (var clusterId in clusterPoints) {
        if (clusterPoints[clusterId].length < 2)
            continue;

        if (clusterPoints[clusterId].length == 2) {
            newCellsDict[clusterId] = clusterPoints[clusterId];
            continue;
        }


        var mesh = d3.geom.delaunay(clusterPoints[clusterId]).filter(function(t) {
            return  (haversine(t[0][0], t[0][1], t[1][0], t[1][1]) < distance &&
                     haversine(t[0][0], t[0][1], t[2][0], t[2][1]) < distance &&
                     haversine(t[2][0], t[2][1], t[1][0], t[1][1]) < distance);
        });

        var collection = {type: "FeatureCollection", features: mesh.map(function(d) {
            return { "type": "Feature",
                    "geometry": {
                        "type": "Polygon",
                        "coordinates": [d]
                    }
                };
            })
        };

        var topology = topojson.topology({collection: collection}, {"verbose": "truthy"});
        console.log('topology', topology);

        /*
        for (var i = 0; i < topology.arcs.length; i++)
            console.log('arc:', topology.arcs[i]);

        for (var i = 0; i < topology.objects.collection.geometries.length; i++)
            console.log('geom:', topology.objects.collection.geometries[i]);
            */

        var merged =  topojson.merge(topology, topology.objects.collection.geometries);
        mergedCoords = merged.coordinates.filter(function(d) { return d.length > 0; })
        .map(function(d) { return d[0]; });
        console.log('mergedCoords:', mergedCoords);

        newCellsDict[clusterId] = mergedCoords;
    }

    return cellsDict;
}

var main = function() {
    var argv = require('minimist')(process.argv.slice(2));

    if (argv._.length < 1) {
        console.log('Not enough arguments:', argv._);
    }

    if ('d' in argv)
        distance = argv['d'];
    else
        distance = 1;

    var fs = require('fs');
    var xml2js = require('xml2js');

    var parser = new xml2js.Parser();
    fs.readFile(argv._[0], function(err, data) {
        parser.parseString(data, function (err, result) {
            var clusterPoints = clusterOsmPoints(result, distance);
            var alphaShapes = clustersToAlphaShapes(clusterPoints, distance);

            var outputStruct = {"bounds": alphaShapes,
                                "distance": distance };

            //console.log(JSON.stringify(outputStruct));
        });
    });
};

if (require.main === module) {
    main();
}

File: scripts/cluster_by_distances.py
#!/usr/bin/python

import numpy as np
import sys
from haversine import haversine
from optparse import OptionParser

import xml.etree.ElementTree as ET

def proximity_cluster(points, distance,dist):
    #cluster points such that whenever two points are
    #within distance of each other, they are placed in
    #the same cluster

    #take m n-d points as input and
    #return an array of length m containing integers
    #where each integers encodes the cluster that the ith
    #point is in
   # 
    # dist is a function that computes the distance between
    # any two points

    print >>sys.stderr, "distance:", distance

    clusters = []
    links = []
    linksDict = {}
    
    # all points are in their own cluster
    for i in range(len(points)):
        clusters.append(i);
        linksDict[i] = [];

    # sort the points by x values
    points.sort(key=lambda x: x[0])

    for i in range(len(points)):
        if i % 100 == 0:
            print "\r{} of {}".format(i, len(points))

        for j in range(i+1, len(points)):
            d1 = dist([points[j][0], points[j][1]],[points[i][0],points[j][1]])
            if d1 > distance:
                break;  # gone too far

            d = dist(points[i], points[j])
            if (dist(points[i], points[j]) < distance):
                # the two points are close enough to be linked
                links.append([i,j])
                linksDict[i].append(j)
                linksDict[j].append(i)

    visited = set()

    queue = []
    # traverse the list of links and assign clusters
    def bfs():
        to_visit = set()
        while len(queue) > 0:
            currNode, prevNode = queue.pop(0)
            if currNode in visited:
                continue
            
            visited.add(currNode);
            clusters[currNode] = min(clusters[currNode], clusters[prevNode]);

            for i in range(len(linksDict[currNode])):
                if linksDict[currNode][i] in visited or linksDict[currNode][i] in to_visit:
                    continue;

                to_visit.add(linksDict[currNode][i])
                queue.append((linksDict[currNode][i], currNode))

    # go through every point and traverse its list of links
    # assigning clusters
    for i in range(len(points)):
        if i in visited:
            continue;

        queue = [(i,i)]
        bfs()

    return {"clusters": clusters, "links": links};

def extract_points(tree, distance=1):
    '''
    Get a list of points to use for clustering from the osm
    file. The points will consist of all nodes, as well as
    extra points that are placed along long ways to make sure
    that they are included.

    @param tree: An XML tree structure.
    @return: A list of points.
    '''
    root = tree.getroot()
    points = []

    nodes = {}
    for node in tree.iter('node'):
        points += [[float(node.attrib['lat']), float(node.attrib['lon'])]]
        nodes[node.attrib['id']] = points[-1]
    for way in tree.iter('way'):
        nds = list(way.iter('nd'))
        for i in range(len(nds)-1):
            node_start = nodes[nds[i].attrib['ref']]
            node_end = nodes[nds[i+1].attrib['ref']]

            vec_length = haversine(node_start, node_end)
            #print >>sys.stderr, "vec_length:", vec_length
            vec_length_parts = int(2 + (1+ vec_length / distance))

            lats = np.linspace(node_start[0], node_end[0], vec_length_parts)
            lons = np.linspace(node_start[1], node_end[1], vec_length_parts) 

            #print >>sys.stderr, "len(lats)", len(lats), len(lons)
            #print >>sys.stderr, "zip", zip(lats, lons)

            #print >>sys.stderr, "points:", zip(lats, lons)
            points += zip(lats, lons)

    return points

def distanceBetweenNodes(node1, node2):
    dist = haversine(node1, node2)
    return dist

def main():
    usage = """
    python cluster_by_distances.py file.osm
    """
    num_args= 0
    parser = OptionParser(usage=usage)

    parser.add_option('-d', '--distance', dest='distance', default=1, help="The distance within which to cluster items", type='int')
    #parser.add_option('-u', '--useless', dest='uselesss', default=False, action='store_true', help='Another useless option')

    (options, args) = parser.parse_args()

    if len(args) < num_args:
        parser.print_help()
        sys.exit(1)

    tree = ET.parse(args[0])
    points = extract_points(tree, options.distance)
    ret = proximity_cluster(points, options.distance, distanceBetweenNodes)
    print "ret[clusters]", ret["clusters"]

if __name__ == '__main__':
    
    main()

